[WebGPU] mip level validation is not handled correctly in copyBuffer/TextureToBuffer/Texture

mwyrzykowski opened the pull request:
#### 77d269ed019a0b6906bee408f713827b469d4e9f
<pre>
[WebGPU] mip level validation is not handled correctly in copyBuffer/TextureToBuffer/Texture
<a href="https://bugs.webkit.org/show_bug.cgi?id=254175">https://bugs.webkit.org/show_bug.cgi?id=254175</a>
&lt;radar://106958194&gt;

Reviewed by NOBODY (OOPS!).

The correct sizes were already computed in logicalMiplevelSpecificTextureExtent
but they were not used for validation.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::copyTextureToBuffer):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeTexture):
3D textures with rowBytes &gt; 2048 will assert in Metal in debug
and release.

* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::validateLinearTextureData):
Metal requires the copy size to be a multiple of the block size.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77d269ed019a0b6906bee408f713827b469d4e9f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3636 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3733 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3826 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5062 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3925 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3725 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3151 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3934 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3283 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4886 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1440 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3254 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3311 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3248 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4669 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3689 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2984 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3223 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3254 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3255 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3501 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -590,11 +614,23 @@ static bool validateCopyBufferToTexture(const WGPUImageCopyBuffer& source, const
+            || destination.origin.x + widthForMetal > logicalSize.width
+            || destination.origin.y + heightForMetal > logicalSize.height
+            || destination.origin.z + depthForMetal > logicalSize.depthOrArrayLayers
+            || sourceBytesPerRow > 2048) {
```
maybe 2048 should be from one of the device limits?

djg replied:
If you don't find a device limit, please document with a FIXME to come back and fix it up later.

djg reviewed (approved):
:shipit:

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -734,6 +779,14 @@ static bool validateCopyTextureToBuffer(const WGPUImageCopyTexture& source, cons
+        if (!widthForMetal || !heightForMetal)
+            return;
+
+        if (widthForMetal * heightForMetal > destinationBuffer.length) {
```
I think this is wrong, at least it causes a regression in compressed texture formats copyTextureToTexture cases

mwyrzykowski committed:
[WebGPU] mip level validation is not handled correctly in copyBuffer/TextureToBuffer/Texture
https://bugs.webkit.org/show_bug.cgi?id=254175
<radar://106958194>

Reviewed by NOBODY (OOPS!).

The correct sizes were already computed in logicalMiplevelSpecificTextureExtent
but they were not used for validation.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::copyTextureToBuffer):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeTexture):
3D textures with rowBytes > 2048 will assert in Metal in debug
and release.

* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::validateLinearTextureData):
Metal requires the copy size to be a multiple of the block size.
